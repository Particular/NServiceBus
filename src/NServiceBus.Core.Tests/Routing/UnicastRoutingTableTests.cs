namespace NServiceBus.Core.Tests.Routing
{
    using System.Collections.Generic;
    using NServiceBus.Routing;
    using NUnit.Framework;

    [TestFixture]
    class UnicastRoutingTableTests
    {
        [Test]
        public void When_group_does_not_exist_routes_are_added()
        {
            var routingTable = new UnicastRoutingTable();
            var route = UnicastRoute.CreateFromEndpointName("Endpoint1");
            routingTable.AddOrReplaceRoutes("key", new List<RouteTableEntry>()
            {
                new RouteTableEntry(typeof(Command), route),
            });

            var retrievedRoute = routingTable.GetRouteFor(typeof(Command));
            Assert.AreSame(route, retrievedRoute);
        }

        [Test]
        public void When_group_exists_routes_are_replaced()
        {
            var routingTable = new UnicastRoutingTable();
            var oldRoute = UnicastRoute.CreateFromEndpointName("Endpoint1");
            var newRoute = UnicastRoute.CreateFromEndpointName("Endpoint2");
            routingTable.AddOrReplaceRoutes("key", new List<RouteTableEntry>()
            {
                new RouteTableEntry(typeof(Command), oldRoute),
            });

            routingTable.AddOrReplaceRoutes("key", new List<RouteTableEntry>()
            {
                new RouteTableEntry(typeof(Command), newRoute),
            });

            var retrievedRoute = routingTable.GetRouteFor(typeof(Command));
            Assert.AreSame(newRoute, retrievedRoute);
        }

        [Test]
        public void When_routes_are_ambiguous_it_throws_exception()
        {
            var routingTable = new UnicastRoutingTable();
            var lowPriorityRoute = UnicastRoute.CreateFromEndpointName("Endpoint1");
            var highPriorityRoute = UnicastRoute.CreateFromEndpointName("Endpoint2");

            routingTable.AddOrReplaceRoutes("key2", new List<RouteTableEntry>()
            {
                new RouteTableEntry(typeof(Command), highPriorityRoute),
            });

            Assert.That(() =>
            {
                routingTable.AddOrReplaceRoutes("key1", new List<RouteTableEntry>()
                {
                    new RouteTableEntry(typeof(Command), lowPriorityRoute),
                });
            }, Throws.Exception);
        }

        class Command
        {
        }
    }
}