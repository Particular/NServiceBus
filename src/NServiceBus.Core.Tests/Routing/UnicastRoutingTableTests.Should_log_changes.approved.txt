Route table refreshed.
NServiceBus.Core.Tests.Routing.UnicastRoutingTableTests+Command -> Endpoint
NServiceBus.Core.Tests.Routing.UnicastRoutingTableTests+Command2 -> [Endpoint-XYZ]
NServiceBus.Core.Tests.Routing.UnicastRoutingTableTests+Command3 -> <Endpoint@Machine>